src/canvas.c

```diff
@@ -466,6 +466,31 @@ wdTranslateWorld(WD_HCANVAS hCanvas, float dx, float dy)
     }
 }
 
+void
+wdSetWorldTransform(WD_HCANVAS hCanvas, const WD_MATRIX* pMatrix)
+{
+    if(d2d_enabled()) {
+        d2d_canvas_t* c = (d2d_canvas_t*) hCanvas;
+        dummy_D2D1_MATRIX_3X2_F m;
+
+        m._11 = pMatrix->m11;
+        m._12 = pMatrix->m12;
+        m._21 = pMatrix->m21;
+        m._22 = pMatrix->m22;
+        m._31 = pMatrix->dx;
+        m._32 = pMatrix->dy;
+        d2d_apply_transform(c, &m);
+    } else {
+        gdix_canvas_t* c = (gdix_canvas_t*) hCanvas;
+
+        dummy_GpMatrix* matrix;
+        gdix_vtable->fn_CreateMatrix2(pMatrix->m11, pMatrix->m12,
```

Comment:
1. You are calling `GdipCreateMatrix2()` but the matrix objects seems to leak. You also need to call `GdipDeleteMatrix()` when you are done with it.

2. Most (if not all) WinDrawLib painting functions do not return anything (`void` return type). Because in my eyes, apps have no reasonable error handling for such functions but just still paint at least something. But we should still never cause correctly written app to e.g. crash, even if some GDI+ or Direct2D function fails. Please add some test whether the  `GdipCreateMatrix2()` succeeds. Passing `NULL` or invalid pointer into the other function below simply can not be allowed.

    In the case of matrix creation failure, just some debug message with `WD_TRACE()` and early `return` would be enough. Look to other code which creates any GDI+ object.